projects/client/RabbitMQ.Client/src/client/impl/AsyncConsumerWorkService.cs

```diff
@@ -43,14 +43,16 @@ class WorkPool
             readonly ConcurrentQueue<Work> _workQueue;
             readonly CancellationTokenSource _tokenSource;
             readonly ModelBase _model;
-            readonly SemaphoreSlim _semaphore = new SemaphoreSlim(0);
+            CancellationTokenRegistration _tokenRegistration;
+            volatile TaskCompletionSource<bool> _syncSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
Pretty sure the volatile is not needed due to the full fence semantics of the async statemachine. Happy to remove if others agree with my reasoning

Reply:
The `volatile` is not needed here even without awaits:

```csharp
public void Enqueue(Work work)
{
    _workQueue.Enqueue(work);
    _syncSource.TrySetResult(true); // there is nothing to re-order in here, because `TrySetResult` uses the value of the reference and cannot be re-ordered
}
```

```csharp
await _syncSource.Task.ConfigureAwait(false); // read with nothing to re-order after it
_syncSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously); // write with nothing to re-order before it
```

IMHO, the only possible issue is not getting the new TCS from the enqueue calling thread.

Reply:
In that case it would not unleash the next iteration and therefore be a bug without the volatile, right? This is one if the things I always have to think multiple times and in the end I chickened out and left it in there

Reply:
Yeap, without `volatile` in the rare scenario where only one thread is filling the queue in a loop, it can always stay with the very first instance of completion source (in theory).
```csharp
while(true)
{
    _ = _syncSource.TrySetResult(true); // <- _syncSource is not changed by this thread so there is no point in re-fetching it from the memory 
}
```
As far as I can see from the code, `TrySetResult` calls `Task.TrySetResult`, which calls `ContingentProperties.SetCompleted()`, which calls `ManualResetEventSlim.Set()`, which calls `Set(false)`, which uses `lock`, which uses `Monitor`, which (didn't check right now, but if I remember correctly) inserts memory barriers. Oh... =) So I guess the code will hold even in this scenario.
I understand, it's really a hard task to track the behavior of the multi-threaded code when you write it. Sorry to bother you, I'm really sorry because the only reason I've got here was my colleague coming to me in the morning: "Hey, I've seen RMQ provides async interfaces, I've looked through the code and I'd like to know why they use volatile."
I must say that I'm also not 100% sure that nothing will break if `volatile` is removed, but I'm 99% sure.

Reply:
No worries, that's why I left the volatile in there for now. I'm also almost sure that it can be removed but only almost so I figured it is probably safer to have it in since it doesn't seem to impact the perf that much.